Userland/Libraries/LibGL/SoftwareGLContext.cpp

```diff
@@ -648,9 +704,101 @@ void SoftwareGLContext::gl_cull_face(GLenum cull_mode)
     m_culled_sides = cull_mode;
 }
 
+GLuint SoftwareGLContext::gl_gen_lists(GLsizei range)
+{
+    if (range <= 0) {
+        m_error = GL_INVALID_VALUE;
+        return 0;
+    }
+    if (m_in_draw_state) {
+        m_error = GL_INVALID_OPERATION;
+        return 0;
+    }
+
+    auto initial_entry = m_listings.size();
+    m_listings.resize(range + initial_entry);
+    return initial_entry + 1;
+}
+
+void SoftwareGLContext::gl_call_list(GLuint list)
+{
+    if (m_gl_call_depth > max_allowed_gl_call_depth)
+        return;
+
+    APPEND_TO_CALL_LIST_AND_RETURN_IF_NEEDED(gl_call_list, list);
+
+    if (m_listings.size() < list)
+        return;
+
+    TemporaryChange { m_gl_call_depth, m_gl_call_depth + 1 };
+
+    auto& listing = m_listings[list - 1];
+    for (auto& entry : listing.entries) {
+        entry.function.visit([&](auto& function) {
+            entry.arguments.visit([&](auto& arguments) {
+                auto apply = [&]<typename... Args>(Args && ... args)
+                {
+                    if constexpr (requires { (this->*function)(forward<Args>(args)...); })
+                        (this->*function)(forward<Args>(args)...);
+                };
+
+                arguments.apply_as_args(apply);
+            });
+        });
+    }
+}
+
+void SoftwareGLContext::gl_delete_lists(GLuint list, GLsizei range)
+{
+    if (m_listings.size() < list || m_listings.size() <= list + range)
+        return;
+
+    for (auto& entry : m_listings.span().slice(list - 1, range))
+        entry.entries.clear();
+}
+
+void SoftwareGLContext::gl_end_list()
+{
```

Comment:
I'm not sure about this
on one hand, the docs don't say this isn't allowed, on the other hand, if this is in `GL_COMPILE` mode, it won't run for the actual listing, and it won't end the listing at all...

Reply:
Looking a little bit more carefully at the docs, `glNewList` cannot be called inside a listing, and `glEndList` always ends a list, so it's not added to the listing either.
Seems like the only listing ops that are recorded are just the glCallList*s.

Reply:
Resolved this by just excluding `glNewList` and `glEndList` from the call list.